registry/client/endpoint.go

```diff
@@ -0,0 +1,268 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/registry/api/v2"
+)
+
+// Authorizer is used to apply Authorization to an HTTP request
+type Authorizer interface {
+	// Authorizer updates an HTTP request with the needed authorization
+	Authorize(req *http.Request) error
```

Comment:
I know we are still working on this, but I really don't want to bring this model from the engine into distribution. It creates a lot of problems with the code. Are we still planning on following the model in the goauth2 library?

Effectively, we want the following stack:

```
| Client      |
---------------
| Session   |
----------------
| Transport |
```

The _Transport_ adds the authentication info to requests (possibly with a `TokenSource` or other kind of authorization mechanism). The `Session` manages the user and endpoint data, possible injecting the transport into an `(*http.Client)`. The `Client` is the application level client that does http operations with an `(*http.Client)` without having to know anything about the `Session` or the `Transport`.

Right now, the main issue with this approach is that endpoint ends up with too many responsibilities.